vent/io/devices/sensors.py

```diff
@@ -14,76 +15,85 @@ class Sensor(ABC):
+        if len(self._data['value']) == 0:
+            self.update()
+        if average:
+            value = sum(self._data['value'])/len(self._data['value'])
+        else:
+            value = self._data['value'].pop()
+        self._clear()
+        return value
 
     def age(self) -> float:
         """ Returns the time in seconds since the last sensor update, or -1 if never updated."""
-        if self._last_timestamp == -1:
+        if not self._data['timestamp']:
             return -1.0
         else:
-            return time.time() - self._last_timestamp
+            return time.time() - self._data['timestamp'][-1]
 
     def reset(self):
         """ Resets the sensors internal memory. May be overloaded by subclasses to extend functionality specific to a
         device.
         """
-        self._data = np.zeros(self.data_length, dtype=np.float16)
-        self._i = 0
+        self._clear()
+            
+    def _clear(self):
+        """ Resets the sensors internal memory. """
+        for key in self._data.keys():
+            self._data[key].clear()
 
     @property
-    def data(self) -> np.ndarray:
-        """ Locally-stored observations.
-        Note: ndarray.astype(bool) returns an equivalent sized array
-        with True for each nonzero element and False everywhere else.
+    def data(self) -> np.array:
+        """ Returns all Locally-stored observations. 
 
         Returns:
-            np.ndarray: an ndarray of observations arranged oldest to newest. Result has length equal to the lessor
-                of self.n and the number of observations.
-        made.
+            np.array: An array of timestamped observations arranged oldest to newest.
         """
-        rolled = np.roll(self._data, self.data_length - self._i)
-        return rolled[rolled.astype(bool)]
+        result = np.array(
+            [[timestamp, value] for timestamp, value in zip(self._data['timestamp'], self._data['value'])]
+        )
+        self._clear()
+        return result
 
     @property
-    def data_length(self) -> int:
+    def maxlen_data(self) -> int:
         """ Returns the number of observations kept in the Sensor's internal ndarray. Once the ndarray has been filled,
         the sensor begins overwriting the oldest elements of the ndarray with new observations such that the size of the
         internal storage stays constant.
         """
-        return self._data_length
+        return self._maxlen_data
 
-    @data_length.setter
-    def data_length(self, new_data_length):
+    @maxlen_data.setter
+    def maxlen_data(self, new_data_length):
         """ Set a new length for stored observations. Clears existing
         observations and resets.
 
         Args:
             new_data_length (int): The new length of internal observation storage
         """
-        self._data_length = new_data_length
-        self.reset()
+        self._maxlen_data = new_data_length

```

Comment:
would try to do the rest of the operation before setting the private attr, but it's a minor thing. also check that arg is an int